streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java

```diff
@@ -1073,6 +1078,35 @@ int commit(final Collection<Task> tasksToCommit) {
         assertTrue(committed.get());
     }
 
+    @Test
+    public void shouldCommitEarlyIfNeeded() {
+        final long commitInterval = 1000L;
+        final Properties props = configProps(false);
+        props.setProperty(StreamsConfig.STATE_DIR_CONFIG, stateDir);
+        props.setProperty(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, Long.toString(commitInterval));
+
+        final StreamsConfig config = new StreamsConfig(props);
+        final ConsumerGroupMetadata consumerGroupMetadata = mock(ConsumerGroupMetadata.class);
+        when(consumer.groupMetadata()).thenReturn(consumerGroupMetadata);
+        when(consumerGroupMetadata.groupInstanceId()).thenReturn(Optional.empty());
+        final Task runningTask = mock(Task.class);
+        final TaskManager taskManager = mockTaskManagerCommit(runningTask, 2);
+        when(taskManager.needsCommit(anyBoolean())).thenReturn(false);
+
+        final TopologyMetadata topologyMetadata = new TopologyMetadata(internalTopologyBuilder, config);
+        topologyMetadata.buildAndRewriteTopology();
+        thread = buildStreamThread(consumer, taskManager, config, topologyMetadata);
+        thread.setNow(mockTime.milliseconds());
+        thread.maybeCommit();
+
+        when(taskManager.needsCommit(anyBoolean())).thenReturn(true);
+        mockTime.sleep(commitInterval - 10L);
+        thread.setNow(mockTime.milliseconds());
+        thread.maybeCommit();
+
+        verify(taskManager, times(2)).commit(mkSet(runningTask));
```

Comment:
I was wondering why the verification expects 2 invocations. I then understood that when `needsCommit()` returns `false` the commit is done because the second condition of the `if`-statement is `true` (i.e. it is the first commit). The second commit happens because `needsCommit()` returns `true`, but the second condition is `false`. 
This setup is not ideal because it does not test whether `needsCommit()` avoids an early commit when it returns `false`.  So you need a third call to `maybeCommit()`. The first call would just make the first commit. The second call is within the commit interval (i.e., the second condition is `false`) and `needsCommit()` returns `false` and the third call is within the commit interval and `needsCommit()` returns `true`. You should also verify the call to `taskManager.commit()` after each call to `thread.maybeCommit()` to be sure that the call was triggered or not for the correct case.

Reply:
I've added the third case, and I've independently verified the invocations of `commit`. I've used `Mockito.clearInvocations` to ensure that the invocations I'm verifying are linked to the correct case; if you know of a better way to do this, please let me know.